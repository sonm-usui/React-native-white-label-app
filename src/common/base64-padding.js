// @flow

export function addBase64Padding(data: string): string {
  const trail = data.length % 4
  if (trail === 0) {
    return data
  }

  return `${data}${'='.repeat(4 - trail)}`
}

export function removeBase64Padding(data: string): string {
  return data.split('=')[0]
}
