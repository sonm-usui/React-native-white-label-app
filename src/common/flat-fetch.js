// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this function is just a wrapper on top of fetch.
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const acceptHeaders = { Accept: 'application/json;flavor=didcomm-msg' }

export async function flatFetch(
  url: string,
  body?: string,
  headers?: { [string]: any }
): Promise<[null | typeof Error, null | string, null | Response]> {
  const options = {
    mode: 'cors',
    ...(body ? { method: 'POST', body } : {}),
    ...(headers ? headers : {}),
  }
  try {
    const response = await fetch(url, options)
    const responseText = await response.text()
    // request redirected
    if (response && response.status === 302) {
      return [null, null, response]
    }
    if (!response.ok) {
      throw new Error(responseText)
    }
    return [null, responseText, response]
  } catch (e) {
    return [e, null, null]
  }
}
