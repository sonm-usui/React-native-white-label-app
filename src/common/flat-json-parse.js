// @flow

export function flatJsonParse(
  text: string
): [null | typeof Error, null | Object] {
  try {
    return [null, JSON.parse(text)]
  } catch (e) {
    return [e, null]
  }
}
