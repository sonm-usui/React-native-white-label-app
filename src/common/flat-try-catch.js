// @flow

export const flatTryCatch = (fn: (...variableArgs: any[]) => any) => (
  ...args: any[]
): [null | typeof Error, null | any] => {
  try {
    return [null, fn(...args)]
  } catch (e) {
    return [e, null]
  }
}
