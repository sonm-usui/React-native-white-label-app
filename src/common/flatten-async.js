// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this is just a helper function to wrap on top of other async functions
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Here are the problems that this function aims to solve
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const flattenAsync = (
  fn: (...variableArgs: any[]) => Promise<any>
) => async (
  ...args: Array<any>
): Promise<[null | typeof Error, null | any]> => {
  try {
    const response = await fn(...args)
    return [null, response]
  } catch (e) {
    return [e, null]
  }
}
