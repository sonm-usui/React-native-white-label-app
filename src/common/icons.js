// @flow
import React from 'react'
import { Icon } from 'react-native-eva-icons'
import { colors } from './styles/constant'
import { moderateScale } from 'react-native-size-matters'

type IconProps = {
  width?: number,
  height?: number,
  color?: string,
}

type CommonIconProps = {
  name: string,
} & IconProps

export const HOME_ICON = 'home-outline'
export const CONNECTIONS_ICON = 'people-outline'
export const SETTINGS_ICON = 'settings-2-outline'

export const ERROR_ICON = 'alert-circle-outline'
export const CAMERA_ICON = 'camera-outline'
export const CHECK_MARK_ICON = 'checkmark-circle-2-outline'
export const HOME_MENU_ICON = 'menu-outline'
export const ANDROID_BACK_ARROW_ICON = 'arrow-back-outline'
export const IOS_BACK_ARROW_ICON = 'arrow-ios-back-outline'
export const MORE_ICON = 'more-vertical-outline'
export const DELETE_ICON = 'trash-2-outline'
export const CLOSE_ICON = 'close-outline'
export const CHAT_ICON = 'message-square-outline'
export const INFO_ICON = 'info-outline'
export const ARROW_RIGHT_ICON = 'chevron-right-outline'
export const LOCK_ICON = 'lock'
export const SAVE_ICON = 'save-outline'
export const ALERT_ICON = 'alert-circle-outline'
export const ARROW_FORWARD_ICON = 'arrow-ios-forward-outline'
export const CHECKMARK_ICON = 'checkmark-outline'
export const SHARE_ICON = 'share-outline'
export const GRID_ICON = 'grid-outline'

// common icon class is implemented to set default values (except name) for icons
// in that case we should only set icon name and get an icon with default color and size
export const EvaIcon = (props: CommonIconProps) => {
  const { width, height, color } = props

  return (
    <Icon
      {...props}
      width={width ? width : moderateScale(24)}
      height={height ? height : moderateScale(24)}
      fill={color ? color : colors.gray2}
    />
  )
}
