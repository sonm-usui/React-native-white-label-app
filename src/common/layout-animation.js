// @flow

import { LayoutAnimation } from 'react-native'
import { checkIfAnimationToUse } from '../bridge/react-native-cxs/RNCxs'

export function animateLayout() {
  if (!checkIfAnimationToUse()) {
    LayoutAnimation.configureNext(
      LayoutAnimation.create(
        150,
        LayoutAnimation.Types.linear,
        LayoutAnimation.Properties.opacity
      )
    )
  }
}
