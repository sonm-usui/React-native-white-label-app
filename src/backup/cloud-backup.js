// @flow

import React, { Component } from 'react'
import {
  View,
  Image,
  Animated,
  Dimensions,
  ActivityIndicator,
} from 'react-native'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { PanGestureHandler, State } from 'react-native-gesture-handler'

import type { Store } from '../store/type-store'

import { cloudBackupRoute, settingsRoute } from '../common'
import { withStatusBar } from '../components/status-bar/status-bar'

import { Container, CustomView, CustomText, CustomButton } from '../components'
import {
  getCloudBackupStatus,
  getAutoCloudBackupEnabled,
} from '../store/store-selector'
import { connectionHistoryBackedUp } from '../connection-history/connection-history-store'
import { cloudBackupStart, resetCloudBackupStatus } from './backup-store'
import { setAutoCloudBackupEnabled } from './backup-actions'
import { color, colors } from '../common/styles/constant'

import { questionStyles } from './styles'
const appImage = require('../images/cb_app.png')
const downloadImage = require('../images/Group.png')

const successImg = require('../images/Success.png')
const errorImg = require('../images/Sad_Face_Red_Error.png')

import type {
  CloudBackupScreenProps,
  NavigateBackToSettingsType,
} from './type-backup'

import { QuestionScreenHeader } from '../question/components/question-screen-header'

import {
  CLOUD_BACKUP_LOADING,
  CLOUD_BACKUP_COMPLETE,
  CLOUD_BACKUP_FAILURE,
  AUTO_CLOUD_BACKUP_ENABLED,
  WALLET_BACKUP_FAILURE,
} from './type-backup'
import { safeSet, walletSet } from '../services/storage'
import { appName } from '../external-imports'

const { height } = Dimensions.get('window')

// TODO: this is copy and pasted questionStyles from the question modal, should refactor once generic modal is made
export class CloudBackup extends Component<CloudBackupScreenProps, void> {
  _translateY = new Animated.Value(0)

  componentDidMount = () => {
    this.props.resetCloudBackupStatus()

    // NOTE: automatically start backing up and show QuestionLoader
    if (this.props.isAutoBackupEnabled) {
      this.props.cloudBackupStart()
    }
    // NOTE: might switxh to automatically start backing up and show QuestionLoader
    // const {fromToggleAction, switchState} = this.props.route.params
    // if (this.props.isAutoBackupEnabled && !fromToggleAction) {
    //   this.props.cloudBackupStart()
    // } else if(fromToggleAction && !switchState) {
    //   walletSet(AUTO_CLOUD_BACKUP_ENABLED, 'false')
    //   this.props.setAutoCloudBackupEnabled(false)
    //   this.props.cloudBackupStart()
    //}
  }

  startCloudBackup = (autoCloudBackupEnabled: boolean) => {
    walletSet(AUTO_CLOUD_BACKUP_ENABLED, autoCloudBackupEnabled.toString())
    safeSet(AUTO_CLOUD_BACKUP_ENABLED, autoCloudBackupEnabled.toString())
    this.props.setAutoCloudBackupEnabled(autoCloudBackupEnabled)
    this.props.cloudBackupStart()
  }

  _getTransform = (translateY) => [{ translateY }]

  _getOpacity = (translateY) =>
    translateY.interpolate({
      inputRange: [0, height],
      outputRange: [1, 0.2],
      extrapolate: 'clamp',
    })

  // this a common pattern when we want to be sure that we are not running
  // actions after a component is unmounted
  // in our case, user can close modal either by pressing "okay" or clicking outside
  // so we need to be sure that we are not running code after this component
  // is unmounted from react-native tree
  isUnmounted = false
  // this variable is just to ensure that on slow devices
  // if user clicks on gray area and component was already scheduled to close
  // by auto close, or by user clicking on okay button and then auto-close trigger
  // we want to make sure that close is triggered by either action
  // and we don't want to run close again in any case
  isCloseTriggered = false

  componentWillUnmount() {
    this.isUnmounted = true
  }
  onCancel = () => {
    !this.isUnmounted &&
      !this.isCloseTriggered &&
      this.props.navigation.navigate(settingsRoute)
    this.isCloseTriggered = true
  }

  _onPanGestureEvent = Animated.event(
    [
      {
        nativeEvent: {
          translationY: this._translateY,
        },
      },
    ],
    {
      useNativeDriver: true,
    }
  )

  _onHandlerStateChange = (event) => {
    const { state, velocityY, translationY } = event.nativeEvent
    if (state === State.END) {
      const minimumDistanceToClose = 150
      if (velocityY > 60 || translationY > minimumDistanceToClose) {
        this.onCancel()
        return
      }

      Animated.spring(this._translateY, {
        toValue: 0,
        useNativeDriver: true,
      }).start()
    }
  }

  getModalBody = (cloudBackupStatus: any) => {
    if (cloudBackupStatus === CLOUD_BACKUP_LOADING) {
      return <QuestionLoader />
    } else if (cloudBackupStatus === CLOUD_BACKUP_COMPLETE) {
      this.props.connectionHistoryBackedUp()
      return <Success navigateBackToSettings={this.navigateBackToSettings} />
    } else if (
      cloudBackupStatus === CLOUD_BACKUP_FAILURE ||
      cloudBackupStatus === WALLET_BACKUP_FAILURE
    ) {
      return <Error navigateBackToSettings={this.navigateBackToSettings} />
    } else {
      return this.mainBody()
    }
  }

  navigateBackToSettings = () => {
    this.props.navigation.navigate(settingsRoute)
  }

  mainBody = () => {
    return (
      <View>
        <CustomView row center>
          <Image
            style={[questionStyles.backupLogo]}
            source={downloadImage}
            resizeMode="cover"
          />
        </CustomView>
        <CustomView row center />
        <CustomText
          center
          size={23}
          numberOfLines={2}
          bg={false}
          bold
          style={[questionStyles.customTextHelperStyles]}
        >
          Enable Automatic Backups?
        </CustomText>
        <CustomText
          size="p"
          bg={false}
          style={[{ color: color.bg.tertiary.font.seventh }]}
        >
          When enabled, any changes to your {appName} app are automatically
          backed up to the Evernym Cloud. This makes recovery easier should you
          lose your phone.
        </CustomText>
        <CustomView style={questionStyles.customViewHelperStyles}>
          <CustomText
            size="p"
            bg={false}
            style={[{ color: color.bg.tertiary.font.seventh }]}
          >
            All {appName} backups are encrypted and anonymized before leaving
            your phone. You will still need your Recovery Phrase and a fresh
            install of {appName} to recover.
          </CustomText>
        </CustomView>

        <CustomView>
          <CustomView row style={[questionStyles.questionActionContainer]}>
            <Container>
              <CustomButton
                twelfth
                style={[
                  questionStyles.actionButton,
                  questionStyles.cancelButton,
                ]}
                title={'Just One Cloud Backup'}
                // testID={`${testID}-cancel`}
                onPress={() => this.startCloudBackup(false)}
              />

              <CustomView style={questionStyles.customViewHelperStyles}>
                <CustomButton
                  primary
                  style={[
                    questionStyles.actionButton,
                    questionStyles.submitButton,
                  ]}
                  title={'Enable Automatic Backups'}
                  // testID={`${testID}-submit`}
                  onPress={() => this.startCloudBackup(true)}
                />
              </CustomView>
            </Container>
          </CustomView>
        </CustomView>
      </View>
    )
  }

  render() {
    const transform = this._getTransform(this._translateY)
    const opacity = this._getOpacity(this._translateY)
    return (
      <Animated.View
        style={[
          questionStyles.container,
          questionStyles.mainContainer,
          {
            opacity,
          },
        ]}
      >
        <Animated.View style={[questionStyles.container, { transform }]}>
          <Container style={[questionStyles.headerContainer]}>
            <PanGestureHandler
              onGestureEvent={this._onPanGestureEvent}
              onHandlerStateChange={this._onHandlerStateChange}
            >
              <Animated.View style={[questionStyles.container]}>
                <QuestionScreenHeader onCancel={this.onCancel} />
              </Animated.View>
            </PanGestureHandler>
          </Container>
          <Animated.View style={[questionStyles.screenContainer]}>
            <View>
              <CustomView
                row
                style={[questionStyles.questionSenderContainer]}
                center
              >
                <CustomView>
                  <Image
                    style={[questionStyles.questionSenderLogo]}
                    source={appImage}
                    resizeMode="cover"
                  />
                </CustomView>
                <Container style={[questionStyles.questionSenderName]}>
                  <CustomText
                    size="h5"
                    numberOfLines={2}
                    bg={false}
                    bold
                    style={[{ color: color.bg.tertiary.font.seventh }]}
                  >
                    Cloud Backup
                  </CustomText>
                </Container>
              </CustomView>
              {this.getModalBody(this.props.cloudBackupStatus)}
            </View>
          </Animated.View>
        </Animated.View>
      </Animated.View>
    )
  }
}

function QuestionLoader() {
  return (
    <CustomView bg="tertiary" style={[questionStyles.questionLoaderContainer]}>
      <ActivityIndicator size="large" color={colors.gray3} />
      <CustomText center bg={false}>
        Creating Backup
      </CustomText>
    </CustomView>
  )
}
function Error({ navigateBackToSettings }: NavigateBackToSettingsType) {
  return (
    <CustomView bg="tertiary" style={[questionStyles.questionLoaderContainer]}>
      <Image
        style={[questionStyles.successErrorImg]}
        source={errorImg}
        resizeMode="cover"
      />
      <CustomText center bg={false}>
        Error creating backup!
      </CustomText>
      <CustomView style={questionStyles.dismissButton}>
        <CustomButton
          primary
          style={[questionStyles.actionButton, questionStyles.errorButton]}
          title={'Done'}
          onPress={navigateBackToSettings}
        />
      </CustomView>
    </CustomView>
  )
}
function Success({ navigateBackToSettings }: NavigateBackToSettingsType) {
  return (
    <CustomView bg="tertiary" style={[questionStyles.questionLoaderContainer]}>
      <Image
        style={[questionStyles.successErrorImg]}
        source={successImg}
        resizeMode="cover"
      />

      <CustomText center bg={false}>
        Successfully backed up
      </CustomText>
      <CustomView style={questionStyles.dismissButton}>
        <CustomButton
          primary
          style={[questionStyles.actionButton, questionStyles.submitButton]}
          title={'Done'}
          onPress={navigateBackToSettings}
        />
      </CustomView>
    </CustomView>
  )
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      cloudBackupStart,
      resetCloudBackupStatus,
      setAutoCloudBackupEnabled,
      connectionHistoryBackedUp,
    },
    dispatch
  )

const mapStateToProps = (state: Store) => {
  return {
    cloudBackupStatus: getCloudBackupStatus(state),
    isAutoBackupEnabled: getAutoCloudBackupEnabled(state),
  }
}

export const cloudBackupScreen = {
  routeName: cloudBackupRoute,
  screen: withStatusBar({ color: colors.white })(
    connect(mapStateToProps, mapDispatchToProps)(CloudBackup)
  ),
}
