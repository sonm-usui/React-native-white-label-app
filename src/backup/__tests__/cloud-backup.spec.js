// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import { CloudBackup } from '../cloud-backup'
import { getNavigation } from '../../../__mocks__/static-data'
import { cloudBackupRoute } from '../../common'

describe('<CloudBackup />', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
  })

  const navigation = getNavigation({
    initialRoute: cloudBackupRoute,
  })

  const restoreStore = {
    status: 'ZIP_FILE_SELECTED',
    error: {
      code: 'Error Code',
      message: 'Error Message',
    },
    restoreFile: {
      fileName: 'File Name',
      fileSize: 100,
      type: 'File Type',
      uri: 'uri',
    },
  }

  function props() {
    return {
      navigation: navigation,
      resetCloudBackupStatus: jest.fn(),
      cloudBackup: jest.fn(),
      cloudBackupStatus: jest.fn(),
      error: 'error message',
      message: 'message',
      restore: restoreStore,
      route: 'restoreWaitRoute',
      isCloudBackupEnabled: false,
      cloudBackupStart: jest.fn(),
      saveFileToAppDirectory: jest.fn(),
      setAutoCloudBackupEnabled: jest.fn(),
      updateStatusBarTheme: jest.fn(),
      connectionHistoryBackedUp: jest.fn(),
    }
  }

  it('should match snapshot', () => {
    const tree = renderer.create(<CloudBackup {...props()} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
