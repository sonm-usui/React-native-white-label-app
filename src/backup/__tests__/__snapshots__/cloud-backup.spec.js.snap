// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CloudBackup /> should match snapshot 1`] = `
<View
  collapsable={false}
  nativeID="animatedComponent"
  style={
    Object {
      "backgroundColor": "rgba(0, 0, 0, 0.8)",
      "flex": 1,
      "opacity": 1,
    }
  }
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "flex": 1,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
        }
      }
    >
      <View
        onGestureEvent={[Function]}
        onHandlerStateChange={[Function]}
      >
        <View
          collapsable={false}
          nativeID="animatedComponent"
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            row={true}
            style={
              Object {
                "backgroundColor": "transparent",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 20,
                  "left": 20,
                  "right": 20,
                  "top": 20,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                }
              }
            />
            <View
              style={
                Object {
                  "justifyContent": "flex-end",
                  "marginBottom": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "backgroundColor": "#F2F2F2",
                    "borderRadius": 6,
                    "height": 6,
                    "width": 51,
                  }
                }
              />
            </View>
            <View
              accessible={true}
              focusable={true}
              hitSlop={
                Object {
                  "bottom": 20,
                  "left": 20,
                  "right": 20,
                  "top": 20,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      nativeID="animatedComponent"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "paddingBottom": 10,
          "paddingLeft": "5%",
          "paddingRight": "5%",
        }
      }
    >
      <View>
        <View
          center={true}
          row={true}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "maxHeight": 90,
              "minHeight": 64,
            }
          }
        >
          <View
            style={Object {}}
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "default": "RelativeImageStub",
                }
              }
              style={
                Array [
                  Object {
                    "height": 32,
                    "width": 32,
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "marginLeft": "5%",
              }
            }
          >
            <Text
              allowFontScaling={false}
              numberOfLines={2}
              style={
                Array [
                  Object {
                    "fontSize": 17,
                  },
                  Object {
                    "fontFamily": "System",
                  },
                  Object {},
                  Object {
                    "fontWeight": "bold",
                  },
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  Object {
                    "color": "#505050",
                  },
                ]
              }
              suppressHighlighting={true}
            >
              Cloud Backup
            </Text>
          </View>
        </View>
        <View>
          <View
            center={true}
            row={true}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <Image
              resizeMode="cover"
              source={
                Object {
                  "default": "RelativeImageStub",
                }
              }
              style={
                Array [
                  Object {
                    "height": 54,
                    "width": 80,
                  },
                ]
              }
            />
          </View>
          <View
            center={true}
            row={true}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          />
          <Text
            allowFontScaling={false}
            numberOfLines={2}
            style={
              Array [
                undefined,
                Object {
                  "fontFamily": "System",
                },
                Object {},
                Object {
                  "fontWeight": "bold",
                },
                Object {
                  "textAlign": "center",
                },
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                Object {
                  "color": "#505050",
                  "fontSize": 23,
                  "marginVertical": 20,
                },
              ]
            }
            suppressHighlighting={true}
          >
            Enable Automatic Backups?
          </Text>
          <Text
            allowFontScaling={false}
            style={
              Array [
                undefined,
                Object {
                  "fontFamily": "System",
                },
                Object {},
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                Object {
                  "color": "#505050",
                },
              ]
            }
            suppressHighlighting={true}
          >
            When enabled, any changes to your 
            AppName
             app are automatically backed up to the Evernym Cloud. This makes recovery easier should you lose your phone.
          </Text>
          <View
            style={
              Object {
                "marginTop": 10,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  undefined,
                  Object {
                    "fontFamily": "System",
                  },
                  Object {},
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  null,
                  Object {
                    "color": "#505050",
                  },
                ]
              }
              suppressHighlighting={true}
            >
              All 
              AppName
               backups are encrypted and anonymized before leaving your phone. You will still need your Recovery Phrase and a fresh install of 
              AppName
               to recover.
            </Text>
          </View>
          <View
            style={Object {}}
          >
            <View
              row={true}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "space-around",
                  "marginVertical": "3%",
                  "paddingBottom": 20,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "borderRadius": 3,
                      "marginLeft": 0,
                      "marginRight": 0,
                      "overflow": "hidden",
                    }
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "busy": false,
                        "disabled": false,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    nativeID="animatedComponent"
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "borderColor": "#A5A5A5",
                        "borderRadius": 5,
                        "borderWidth": 1,
                        "opacity": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#ffffff",
                          "borderColor": "#A5A5A5",
                          "borderRadius": 5,
                          "borderWidth": 1,
                          "color": "#777777",
                          "flexDirection": "row",
                          "justifyContent": "center",
                          "padding": 8,
                          "textAlign": "center",
                        }
                      }
                    >
                      <Text
                        style={
                          Object {
                            "color": "white",
                            "fontSize": 18,
                            "paddingBottom": 1,
                            "paddingTop": 2,
                            "textAlign": "center",
                          }
                        }
                      >
                        Just One Cloud Backup
                      </Text>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    Object {
                      "marginTop": 10,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "borderRadius": 3,
                        "marginLeft": 0,
                        "marginRight": 0,
                        "overflow": "hidden",
                      }
                    }
                  >
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        Object {
                          "busy": false,
                          "disabled": false,
                        }
                      }
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      nativeID="animatedComponent"
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "backgroundColor": "#86B93B",
                          "borderColor": "#86B93B",
                          "borderRadius": 5,
                          "borderWidth": 1,
                          "opacity": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "backgroundColor": "#777777",
                            "borderColor": "#86B93B",
                            "borderRadius": 5,
                            "borderWidth": 1,
                            "color": "#ffffff",
                            "flexDirection": "row",
                            "justifyContent": "center",
                            "padding": 8,
                            "textAlign": "center",
                          }
                        }
                      >
                        <Text
                          style={
                            Object {
                              "color": "white",
                              "fontSize": 18,
                              "paddingBottom": 1,
                              "paddingTop": 2,
                              "textAlign": "center",
                            }
                          }
                        >
                          Enable Automatic Backups
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
